core/src/main/java/org/jruby/truffle/nodes/rubinius/TimePrimitiveNodes.java

```diff
@@ -0,0 +1,316 @@
+            writeOffsetNode.execute(time, offset);
+            return time;
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "time_seconds")
+    public static abstract class TimeSecondsPrimitiveNode extends RubiniusPrimitiveNode {
+
+        public TimeSecondsPrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public TimeSecondsPrimitiveNode(TimeSecondsPrimitiveNode prev) {
+            super(prev);
+        }
+
+        @Specialization
+        public long timeSeconds(RubyTime time) {
+            return time.getSeconds();
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "time_useconds")
+    public static abstract class TimeUSecondsPrimitiveNode extends RubiniusPrimitiveNode {
+
+        public TimeUSecondsPrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public TimeUSecondsPrimitiveNode(TimeUSecondsPrimitiveNode prev) {
+            super(prev);
+        }
+
+        @Specialization
+        public long timeUSeconds(RubyTime time) {
+            return time.getNanoseconds();
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "time_decompose")
+    public static abstract class TimeDecomposePrimitiveNode extends RubiniusPrimitiveNode {
+
+        @Child protected RubyTimeToDateTimeNode toDateTimeNode;
+
+        public TimeDecomposePrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            toDateTimeNode = new RubyTimeToDateTimeNode(context, sourceSection);
+        }
+
+        public TimeDecomposePrimitiveNode(TimeDecomposePrimitiveNode prev) {
+            super(prev);
+            toDateTimeNode = prev.toDateTimeNode;
+        }
+
+        @Specialization
+        public RubyArray timeDecompose(VirtualFrame frame, RubyTime time) {
+            final DateTime dateTime = toDateTimeNode.toDateTime(frame, time);
+            final Object[] decomposed = decompose(dateTime);
+            return new RubyArray(getContext().getCoreLibrary().getArrayClass(), decomposed, decomposed.length);
+        }
+
+        @CompilerDirectives.TruffleBoundary
+        public Object[] decompose(DateTime dateTime) {
+            final int sec = dateTime.getSecondOfMinute();
+            final int min = dateTime.getMinuteOfDay();
+            final int hour = dateTime.getHourOfDay();
+            final int day = dateTime.getDayOfMonth();
+            final int month = dateTime.getMonthOfYear();
+            final int year = dateTime.getYear();
+            final int wday = dateTime.getDayOfWeek();
+            final int yday = dateTime.getDayOfYear();
+            final Object isdst = getContext().getCoreLibrary().getNilObject();
+            final Object zone = getContext().getCoreLibrary().getNilObject();
+            return new Object[]{sec, min, hour, day, month, year, wday, yday, isdst, zone};
+        }
```

Comment:
That sounds a lot of complexity. Are those calculations really so complicated? We could maybe also just make sure the compiler understands the concept of a function call that does not side-effect and can be dead code eliminated in case its value is not needed.